### PR TITLE
Rate limit WebSocket messages - Closes #3975

### LINKED ISSUE #3975: Rate limit WebSocket messages
**Story**
As a node user, I would like to restrict traffic/ messages which I receive from any peer so that I can be safeguarded from the spamming.

**Acceptance Criteria**
1. Accept messages/ transaction blocks from each peer when it is not exceeded your Rate limit.
2. If a specific peer is reached his maximum Rate limit, then the specific peer connection should be interrupted.
3. The rate limit  (receiving limit) can vary from peer to peer.

### Expected behavior

There should be a simple rate-limiting mechanism to prevent nodes from spamming our WebSocket connection.

### Actual behavior

There is no rate limiting to prevent remote clients from spamming a node's WebSocket connection.

If we have a sufficiently large machine (or multiple machines), we can spam a specific node with WebSocket messages to make it consume CPU.

### PR DESCRIPTION
**Pull request by jondubois**
### What was the problem?

There was no rate limiting at the WebSocket level to block malicious peers.

### How did I fix it?

- Track the raw WebSocket message rate. If a peer tries to send more than `wsMaxMessageRate` messages per second, they will be disconnected with a forbidden message.
- Adjusted how the p2p request and p2p message rate is calculated to be consistent with the calculation of the `wsMaxMessageRate` and to increase precision.

### How to test it?

New integration test cases were added.

`npm run test:integration`

### Review checklist

* The PR resolves #3975
* All new code is covered with unit tests
* All new code was formatted with Prettier
* Linting passes
* Tests pass
* Commit messages follow the [commit guidelines](CONTRIBUTING.md#git-commit-messages)
* Documentation has been added/updated

### CODE REVIEW

File: elements/lisk-p2p/src/peer/base.ts [hunk 1]
```diff
@@ -173,21 +184,53 @@ export class Peer extends EventEmitter {
...
+				this.disconnect(FORBIDDEN_CONNECTION, FORBIDDEN_CONNECTION_REASON);
+
+				return;
+			}
+
+			this._rpcRates = new Map(
+				[...this._rpcCounter.entries()].map(entry => {
+					const rate = entry[1] / this._rateInterval;
+
+					return [entry[0], rate] as any;
```
Comment (ishantiw) [inline elements/lisk-p2p/src/peer/base.ts]: isn't `entry[0]` is string type and we can use it as key for `rpcRates` type as `Map<string, number>`?

File: elements/lisk-p2p/src/p2p.ts [hunk 2]
```diff
@@ -124,6 +124,7 @@ export const DEFAULT_POPULATOR_INTERVAL = 10000;
 export const DEFAULT_SEND_PEER_LIMIT = 25;
 // Max rate of WebSocket messages per second per peer.
 export const DEFAULT_WS_MAX_MESSAGE_RATE = 100;
+export const DEFAULT_WS_MAX_MESSAGE_RATE_PENALTY = 10;
```
Comment (mitsujutsu) [inline elements/lisk-p2p/src/p2p.ts]: Should penalty be `100`? Otherwise the actual number of messages allowed would be 10x

Comment (jondubois) [reply elements/lisk-p2p/src/p2p.ts]: @mitsujutsu `DEFAULT_WS_MAX_MESSAGE_RATE` means the rate above which we disconnect, not ban. Banning is more serious; we want to restrict it to intentional DoS attempts, not one-offs. For example, if someone is running a node which has bad configs or custom logic, we want to give the operator of that node a chance to fix their configs before they are banned from the network.

File: elements/lisk-p2p/test/integration/p2p.ts [hunk 3]
```diff
@@ -1030,6 +976,193 @@ describe('Integration tests for P2P library', () => {
...
+							}
+						}
+						let peerRates = requestRates.get(p2p.nodeInfo.wsPort) || [];
+						peerRates.push(request.rate);
+						requestRates.set(p2p.nodeInfo.wsPort, peerRates);
+					});
+				}
+			});
+
+			it('should track the request rate correctly when receiving requests', async () => {
```
Comment (diego-G) [inline elements/lisk-p2p/test/integration/p2p.ts]: Shall we add another test case to check if the peer interrupts the channel?

File: elements/lisk-p2p/src/p2p.ts (hunk 2 above)
Comment (mitsujutsu) [reply elements/lisk-p2p/src/p2p.ts]: I see. So in practice they are allowed to exceed the max rate 10 times before they are banned?

File: elements/lisk-p2p/test/integration/p2p.ts (hunk 3 above)
Comment (jondubois) [reply elements/lisk-p2p/test/integration/p2p.ts]: @diego-G Ok resolved.

File: elements/lisk-p2p/src/peer/base.ts [hunk 4]
```diff
@@ -175,21 +187,54 @@ export class Peer extends EventEmitter {
...
+
+			if (this._wsMessageRate > this._peerConfig.wsMaxMessageRate) {
+				this.disconnect(FORBIDDEN_CONNECTION, FORBIDDEN_CONNECTION_REASON);
+				this.applyPenalty(this._peerConfig.wsMaxMessageRatePenalty);
+
+				return;
+			}
+
+			this._rpcRates = new Map(
+				[...this._rpcCounter.entries()].map(entry => {
```
Comment (ishantiw) [inline elements/lisk-p2p/src/peer/base.ts]: maybe for clarity we can use `.map([key, rate]) =>` or `.map([key, value]) =>`?

File: elements/lisk-p2p/src/peer/base.ts [hunk 5]
```diff
@@ -175,21 +187,54 @@ export class Peer extends EventEmitter {
...
+					const rate = entry[1] / this._rateInterval;
+
+					return [entry[0], rate] as any;
+				}),
+			);
+			this._rpcCounter = new Map();
+
+			this._messageRates = new Map(
+				[...this._messageCounter.entries()].map(entry => {
```
Comment (ishantiw) [inline elements/lisk-p2p/src/peer/base.ts]: same here

File: elements/lisk-p2p/src/peer/base.ts (hunk 1 above)
Comment (ishantiw) [reply elements/lisk-p2p/src/peer/base.ts]: we don't need to typecast as `any`, it doesn't give me warning

Comment (jondubois) [reply elements/lisk-p2p/src/peer/base.ts]: It gives me warnings in vscode if I don't put `any`.

**Pull request merged**